### PR TITLE
(on getgov-za) Ticket #1851: Show how a user was verified

### LINKED ISSUE #1851: Django Admin: Indicate how a user was verified.
### Issue description

Problem: for research purposes, the analyst wants to know how reliable a creator is. One indicator for this is how they were verified in the system. Currently, the analyst needs to just infer whether they have gone through IAL2 vetting, based on whether their name is present and capitalized in the system. We wish to disambiguate this and be clearer with precisely how the user has been verified.

Solution: Create an indicator showing how the creator was verified. There are three options:

1. User was migrated from previous registrar, and is grandfathered in. Indicator = X for existing.
2. User is New and has gone thru login.gov IAL2 verification. Indicator = V for Verified.
3. User is New and has failed or opted out of IAL2, and has been "Verified by Staff" Indicator = S for Staff

### Acceptance criteria

- [ ] Creator has an indicator by their name, as listed above, showing how they are verified.

### Additional context

see diagram

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1851

## Changes

- Modifies user.py and adds a VerificationTypeChoices textchoice field, as well as a verification_choice property w/ associated setters
- Modifies djangooidc/views.py to set the "verification_type" (which tracks how the user was verified) on login_callback with a special state for fixture users
- Displays the verification type field on the user object, and on whichever page the creator is present on **(readonly only for now, let me know if this should change** - the logic here is that superusers do not need additional information, much like what we do for some of the other info fields)
- Modifies fixtures to set a default verification_type
- Adds unit tests
- Adds a script to populate existing verification type fields (plus instructions)
- Adds a generic mass_populate_field script  template since we've made a few of these already. This will save us time in the future.
- Migrations 

## Context for reviewers
This PR adds a new field to user called "verification_type" which specifies how the user themselves was verified (through OIDC). For instance, a user who exists in the "VerifiedByStaff" stable went through IAL1 verification and was "Verified by staff". This field is then shown next to the associated "creator" field on DomainRequest. 

This field gets set on `login_callback` which gets triggered through djangooidc. 

**Context for design**
In addition to the changes listed in the original ticket, this PR also exposes a field called "Verification type" on the `User`. This means that any analyst can click any user and get information about them, if they so choose.

It is also worth noting that the design next to creator, which showcases the verification type, is slightly different for superusers because they can actually change that field. Its pushed "down", though I wonder if we should show it to them at all?

## Setup
**Warning:** This PR is slightly time intensive to review. All of the steps below must be followed in order (that is, each section must be done in sequence)

#### To review this PR, you will need the following (at minimum):
1. A analyst account
2. A superuser account
3. A new account which you've just created through login.gov (through the sandbox this ticket exists in)

#### Follow these steps to create a new account:
1. Navigate to /admin
2. Click Verified by staff 
3. Add a new entry
3. Set the email to be that of a new account which you will create through login.gov
4. On my sandbox (important), create a new account on login.gov. If you want to make a new email quickly, I recommend using [this service](https://proton.me/mail) since account creation is just email + password.
5. Ensure that this account is authenticated through IAL1 (Email, password, MFA but no ID checks)
6. Check that everything works normally on the main page
7. Create a new domain request
8. Submit the new domain request (recommended)

#### Verify the "Verified by staff" display value:
1. Login as an analyst
2. Navigate to /admin
3. Click Domain request and find your record
4. On the record itself, find the `Creator` field. You should see `(Verified by staff)` next to it
5. Click the user record, and verify that the verification type field matches
6. Repeat 1-5 as a superuser
7. (As a superuser) Delete the domain request you just created
8. (As a superuser) Delete the new user that you just created
9. **Steps 7-8 are vitally important**

#### Verify the "Legacy user" display value:
1. Create a new Transition Domain. For username (or email, but username is easier), set the email of your new account to the one you created. For domain name, set it to an existing domain in our DB.
2. Login with your new account, navigate to the main page
3. Login as a superuser
4. Go to the same domain that you set in the "domain name" field
5. Set the creator to your new account and save (this can be done automatically with scripts, but this is simpler for review purposes)
6. Note the creator field. It should say "Legacy user"
7. Check the underlying User object, and make sure it looks right.
8. Login as an analyst
9. Navigate to /admin
10. Click Domain information and find your record
11. Repeat steps 6-7
12. (As a superuser) Delete the domain request you created
13. (As a superuser) Delete the new user that you created
 
#### Verify the "Invited by a domain manager" display value:
1. On the main page, click "manage" on an existing domain
2. Click Domain managers
3. Click add a domain manager
4. Add the email of the new account you created
5. Login as your new account, verify that you can access this domain
6. Login as an analyst
7. Navigate to /admin
8. Click Domain request for the underlying domain and find your record
9. On the record itself, find the `Creator` field. You should see `(Invited by a domain manager)` next to it
10. Click the user record, and verify that the verification type field matches
6. Repeat 7-10 as a superuser
8. (As a superuser) Delete the domain request you just created
9. (As a superuser) Delete the new user that you just created

#### Verify the "Created by fixtures" display value:
This must be tested locally (otherwise the script itself is untestable). For designers, 
1. As a superuser or analyst, click on any domain request generated by fixtures (randomly generated data)
2. Note the value next to creator `Created by fixtures`
3. Verify that this display is working for the user

#### Verify that the populate script is working
Run `./manage.py populate_verification_type` either on my sandbox. Note that all fixture users should now have the value "Verified by login.gov" which is the default value for most users.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by erinysong** [comment]
Found discrepancy in testing Legacy User steps 8-11. When navigating to domain request for the domain we changed, the creator is still listed as the original creator and not the legacy user we assigned it to. Was curious what the relationship/flow was among the transition domain, domain, and domain request we're testing in these steps. @zandercymatics 

**Deleted then restored user account displays as legacy** 
image

**Domain request for domain still displays creator as original user instead of legacy user we assigned it to**
image

**Comment by erinysong** [comment]
Found a similar discrepancy when testing for a user with **Invited by Domain Manager status.** Wanted to ask how adding a domain manager to an existing, approved domain is expected to impact its domain request. Was curious why adding a user as a domain manager would make that user the creator in Invited by Domain Manager step 9

File: src/djangooidc/tests/test_views.py [hunk 1]
```diff
@@ -228,6 +238,128 @@ def test_login_callback_initializes_client_and_succeeds(self, mock_client):
                         # assert that redirect is to / when no 'next' is set
                         self.assertEqual(response.status_code, 302)
                         self.assertEqual(response.url, "/")
+    
+    @less_console_noise_decorator
+    def test_login_callback_sets_verification_type_regular(self, mock_client):
```
Comment (therealslimhsiehdy) [inline src/djangooidc/tests/test_views.py:243]: **(Very smol nitpick):** Would it be possible to leave a comment here mentioning that REGULAR meaning "Verified by Login.gov"?

File: src/djangooidc/views.py [hunk 2]
```diff
@@ -99,8 +99,22 @@ def login_callback(request):
             return CLIENT.create_authn_request(request.session)
         user = authenticate(request=request, **userinfo)
         if user:
+
+            # Fixture users kind of exist in a superposition of verification types,
+            # because while the system "verified" them, if they login,
+            # we don't know how the user themselves was verified through login.gov until
+            # they actually try logging in. This edge-case only matters in non-production environments.
+            fixture_user = User.VerificationTypeChoices.FIXTURE_USER
+            is_fixture_user = user.verification_type and user.verification_type == fixture_user
```
[the remark this answers is not in the document]
Comment (therealslimhsiehdy) [reply src/djangooidc/views.py:108]: Discussed synchronously in the Slack Huddle, and it makes sense that we want to show a "Verification Unknown" state versus a NULL state (or at least make it more descriptive of what state it's in to Analysts)

File: src/registrar/management/commands/utility/terminal_helper.py [hunk 3]
```diff
@@ -58,6 +59,54 @@ def bulk_update_fields(model_class, update_list, fields_to_update, batch_size=10
...
+        logger.info("Updating...")
+
+        to_update: List[sender] = []
+        failed_to_update: List[sender] = []
+        for updated_object in objects:
+            try:
+                self.populate_field(updated_object)
+                to_update.append(updated_object)
+            except Exception as err:
+                to_update.append(updated_object)
```
Comment (therealslimhsiehdy) [inline src/registrar/management/commands/utility/terminal_helper.py:95]: **Question:** Should this be `failed_to_update` instead of `to_update`

File: src/registrar/management/commands/utility/terminal_helper.py [hunk 4]
```diff
@@ -58,6 +59,54 @@ def bulk_update_fields(model_class, update_list, fields_to_update, batch_size=10
...
+            """,
+            prompt_title="Do you wish to patch this data?",
+        )
+        logger.info("Updating...")
+
+        to_update: List[sender] = []
+        failed_to_update: List[sender] = []
+        for updated_object in objects:
+            try:
+                self.populate_field(updated_object)
```
Comment (erinysong) [inline src/registrar/management/commands/utility/terminal_helper.py:92]: Noticed that populate_field(self, field_to_update) always raises a NotImplementedError and was wondering what the reasoning was behind doing a try/catch for a function that will always raise an exception

File: src/registrar/models/user.py [hunk 5]
```diff
@@ -114,23 +145,61 @@ def needs_identity_verification(cls, email, uuid):
...
-        # in the username field of a TransitionDomain)
-        if TransitionDomain.objects.filter(username=email).exists():
-            return False
+        # We can't set the verification type here because the user may not
+        # always exist at this point. We do it down the line.
+        verification_type = cls.get_verification_type_from_email(email)
+
+        # The user needs identity verification if they don't meet
+        # any special criteria, i.e. we are validating them "regularly"
+        needs_verification = verification_type == cls.VerificationTypeChoices.REGULAR
```
Comment (erinysong) [inline src/registrar/models/user.py:154]: Can this be simplified to something like return verification_type == cls.VerificationTypeChoices.Regular? And we can just comment that this value indicates it needs verification as above?

File: src/registrar/models/user.py [hunk 6]
```diff
@@ -114,23 +145,61 @@ def needs_identity_verification(cls, email, uuid):
...
+        if verification_type == User.VerificationTypeChoices.INVITED:
+            invitation = (
+                DomainInvitation.objects.filter(email=email_or_username, status=retrieved)
+                .order_by("created_at")
+                .first()
+            )
+
+            # If you joined BEFORE the oldest invitation was created, then you were verified normally.
+            # (See logic in get_verification_type_from_email)
+            if invitation is not None and self.date_joined < invitation.created_at:
```
Comment (erinysong) [inline src/registrar/models/user.py:177]: can invitation is not None be simplified to invitation here?

File: src/registrar/tests/test_management_scripts.py [hunk 7]
```diff
@@ -25,6 +26,103 @@
...
+        )
+        self.grandfathered_user, _ = User.objects.get_or_create(username="grandpa@igormail.gov")
+
+        invited, _ = DomainInvitation.objects.get_or_create(
+            email="invited@igormail.gov", domain=self.domain_1, status=DomainInvitation.DomainInvitationStatus.RETRIEVED
+        )
+        self.invited_user, _ = User.objects.get_or_create(username="invited@igormail.gov")
+
+        self.untouched_user, _ = User.objects.get_or_create(
+            username="iaminvincible@igormail.gov", verification_type=User.VerificationTypeChoices.GRANDFATHERED
```
Comment (therealslimhsiehdy) [inline src/registrar/tests/test_management_scripts.py:67]: That username 🤣

**Review by Katherine-Osos — APPROVED** [review]

File: src/registrar/management/commands/utility/terminal_helper.py [hunk 8]
```diff
@@ -58,6 +59,55 @@ def bulk_update_fields(model_class, update_list, fields_to_update, batch_size=10
...
+                logger.error(err)
+                logger.error(f"{TerminalColors.FAIL}" f"Failed to update {updated_object}" f"{TerminalColors.ENDC}")
+
+        # Do a bulk update on the first_ready field
+        ScriptDataHelper.bulk_update_fields(sender, to_update, fields_to_update)
+
+        # Log what happened
+        TerminalHelper.log_script_run_summary(to_update, failed_to_update, skipped=[], debug=True)
+
+    @abstractmethod
```
Comment (therealslimhsiehdy) [inline src/registrar/management/commands/utility/terminal_helper.py:105]: **Question:** Could you ELI5 why an abstractmethod needs to be defined here? I understand the call in `populate_verification_type.py`, but am confuzzled to why it needs to be set here with the `NotImplementedError`

File: src/registrar/assets/sass/_theme/_admin.scss [hunk 9]
```diff
@@ -617,3 +617,8 @@ address.dja-address-contact-list {
 .usa-button__small-text {
     font-size: small;
 }
+
+// Get rid of padding on all help texts
+form .aligned p.help, form .aligned div.help {
+    padding-left: 0px !important;
```
Comment (erinysong) [inline src/registrar/assets/sass/_theme/_admin.scss:623]: nit: Is there an option to remove the !important tag here?

**Review by erinysong — APPROVED** [review]
Reviewed with Rebecca. Conditionally approving after Rebecca's question about the function of having this script inherit from ABC is explained

**Review by therealslimhsiehdy — APPROVED** [review]
Reviewed with Erin: conditional approval with needing a response to Erin's comment + our question regarding the abstraction method. 

Otherwise looks good as we all synced in the dev channel regarding the visual testing changes for each section (Verified by Staff, Legacy User, etc)

File: src/registrar/management/commands/utility/terminal_helper.py (hunk 8 above)
Comment (therealslimhsiehdy) [reply src/registrar/management/commands/utility/terminal_helper.py:105]: Thank you so much for the detailed explanation! We were thinking along the same lines that this would be used as a template, but wanted to make sure ❤️ If it's not a big ask, would it be possible add this explanation/add some documentation if ppl were to use this template later to make their own scripts in `docs/operations/data_migration.md`? 😅
